### PR TITLE
[FIX] stock: do not merge delivery orders for different customers

### PR DESCRIPTION
**Pull request by lase-odoo**
### Issue:

Pickings associated to different customers are merged in 2 steps delivery setups.

### Steps to reproduce:

- Create 2 storable products: P1 and P2
- Enable multistep routes in the settings > set Warehouse Routes
- Create a new warehouse WH2 with 2 steps delivery
- Go to inventory > Pick (WH2)
- Create and validate a new pick for Customer 1: 1 x P1 
Note: A delivery order was created for customer 1
 - Create and validate a new pick for Customer 2: 1 x P2

### Expected behavior:

A delivery order is created for customer 2

### Current behavior:

The new delivery order is merged with delivery order created for customer 1 and the delivery address is erased as the picking would be associated with 2 contacts.

### Cause of the issue:

When generated for the customer delivery is generated by the 2 step delivery the `_assign_picking` methods tries to assign the move to an existing picking if possible:
https://github.com/odoo/odoo/blob/71b4a97c9e3da94bfc0130371db2c02beb81eebe/addons/stock/models/stock_move.py#L1250-L1254 However, the `partner_id` is only used in the search domain of this picking if the usage of the `location_id/dest_id` is a transit: https://github.com/odoo/odoo/blob/71b4a97c9e3da94bfc0130371db2c02beb81eebe/addons/stock/models/stock_move.py#L1240-L1241 This should also be the case when the destination is a customer.

### Note:

The issue did not occur prior to saas-17.2 since no "Out" transfer were generated by manual pick transfer in 2 steps delivery prior to this version.

opw-3857526
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_move.py [hunk 1]
```diff
@@ -621,6 +621,9 @@ def create(self, vals_list):
         for vals in vals_list:
             if (vals.get('quantity') or vals.get('move_line_ids')) and 'lot_ids' in vals:
                 vals.pop('lot_ids')
+            picking_id = self.env['stock.picking'].browse(vals.get('picking_id'))
```
Comment (HANNICHE-Walid) [inline addons/stock/models/stock_move.py:624]: There should be a merge conflict 
I merged this change yesterday on 17.2

File: addons/stock/models/stock_move.py [hunk 2]
```diff
@@ -1237,7 +1237,7 @@ def _search_picking_for_assignation_domain(self):
             ('picking_type_id', '=', self.picking_type_id.id),
             ('printed', '=', False),
             ('state', 'in', ['draft', 'confirmed', 'waiting', 'partially_available', 'assigned'])]
-        if self.partner_id and (self.location_id.usage == 'transit' or self.location_dest_id.usage == 'transit'):
+        if self.partner_id:
```
Comment (HANNICHE-Walid) [inline addons/stock/models/stock_move.py:1240]: if this condition is too lax 
try self.partner_id any(loc.usage in ( 'transit', 'supplier', 'customer') for loc in (self.location_id, self.location_dest_id)):

**Review by HANNICHE-Walid — APPROVED** [review]
LGTM after you rebase

**Comment by lase-odoo** [comment]
@Whenrow Thanks for the suggestion ! I have updated the diff accordingly but let me know if I miss understood :slightly_smiling_face:

File: addons/stock/models/stock_move.py [hunk 3]
```diff
@@ -1210,7 +1210,7 @@ def _onchange_product_uom(self):
     def _key_assign_picking(self):
         self.ensure_one()
         keys = (self.group_id, self.location_id, self.location_dest_id, self.picking_type_id)
-        if self.partner_id and (self.location_id.usage == 'transit' or self.location_dest_id.usage == 'transit'):
+        if self.partner_id and not self.group_id and any(loc.usage in ('transit', 'supplier', 'customer') for loc in (self.location_id, self.location_dest_id)):
```
Comment (HANNICHE-Walid) [inline addons/stock/models/stock_move.py:1213]: @lase-odoo I think he meant just  `if self.partner_id and not self.group_id:`

Comment (lase-odoo) [reply addons/stock/models/stock_move.py:1213]: Oh well ^^

**Comment by Whenrow** [comment]
robodoo r+ 
thank you

**Pull request closed without merging**